### PR TITLE
prov/efa: use fi_read to copy data to gpu buffer

### PR DESCRIPTION
**Pull request by wzamazon**
This patch use fi_read instead of cudaMemcpy() to copy data to gpu buffer. This change is because cudaMemcpy()
can cause deadlock, when the EFA provider is used by a NCCL plugin.

Related GitHub issue:
   https://github.com/ofiwg/libfabric/issues/6124

The core change is to extend the function rxr_pkt_proc_data(), which will copy data to recieving buffer for all packets that
may contain data.

Each packet type implements a handler function to handle data copied event.

If recieving buffer is on host memory, rxr_pkt_proc_data() will copy data then call the handler fucntions.

If receiving buffer is on GPU memory, rxr_pkt_proc_data() will post a read, and the progress engine will call the handler functions.

Signed-off-by: Wei Zhang <wzam@amazon.com>

### CODE REVIEW

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
bot:aws:retest

File: prov/efa/src/rxr/rxr_pkt_cmd.c [hunk 1]
```diff
@@ -529,7 +610,13 @@ void rxr_pkt_handle_send_completion(struct rxr_ep *ep, struct fi_cq_data_entry *
 	peer = rxr_ep_get_peer(ep, pkt_entry->addr);
 	if (!peer->is_local)
 		rxr_ep_dec_tx_pending(ep, peer, 0);
-	rxr_pkt_entry_release_tx(ep, pkt_entry);
+
+	/*
+	 * If ptk_entry->state == RXR_PKT_ENTRY_COPY_BY_READ, the
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_pkt_cmd.c:615]: typo here

File: prov/efa/src/rxr/rxr_pkt_cmd.c [hunk 2]
```diff
@@ -529,7 +610,13 @@ void rxr_pkt_handle_send_completion(struct rxr_ep *ep, struct fi_cq_data_entry *
...
+	 * handle_data_copied function will release pkt_entry
+	 */
+	if (pkt_entry->state == RXR_PKT_ENTRY_IN_USE)
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_pkt_cmd.c:618]: What about `RXR_PKT_ENTRY_RNR_RETRANSMIT`? Shouldn't this check just be `!= RXR_PKT_ENTRY_COPY_BY_READ` instead?

File: prov/efa/src/rxr/rxr_pkt_type_data.c [hunk 3]
```diff
@@ -304,16 +297,101 @@ int rxr_pkt_proc_data(struct rxr_ep *ep,
...
+ *
+ * It can be called by either rxr_pkt_handle_data_recv() or
+ * rxr_pkt_handle_readrsp_recv() depend on packet type.
+ *
+ * It will copy data from data pkt_entry to receiving buffer in
+ * rx_entry.
+ *
+ * There can be two cases:
+ *
+ *    1. If the receiving buffer is on host memory, this function will
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_pkt_type_data.c:328]: I thought you said you fixed this? There's no reason to repeat the two cases on each function that does it, this should be documented in one place.

File: prov/efa/src/rxr/rxr_pkt_type_data.c [hunk 4]
```diff
@@ -304,16 +297,101 @@ int rxr_pkt_proc_data(struct rxr_ep *ep,
...
+ *    2. If the receiving buffer is on GPU memory, this function will
+ *       call rxr_pkt_post_read_to_copy_data(), which post a RDMA read
+ *       to copy data.
+ *
+ *       If the read was issued successfully,
+ *       rxr_pkt_handle_eager_rtm_data_copied()
+ *       will be called by rxr_pkt_handle_copy_by_read_completion(),
+ *       when the completion entry of the read was received.
+ *
+ *       If the read failed, an error completion entry will be written
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_pkt_type_data.c:341]: Comments like this aren't needed, it's obvious from the code.

File: prov/efa/src/rxr/rxr_pkt_type_req.c [hunk 5]
```diff
@@ -799,28 +765,130 @@ struct rxr_rx_entry *rxr_pkt_get_tagrtm_rx_entry(struct rxr_ep *ep,
...
+ *
+ *       If the read failed, an error code will be returned
+ */
+static inline
+ssize_t rxr_pkt_proc_matched_eager_rtm(struct rxr_ep *ep,
+				       struct rxr_rx_entry *rx_entry,
+				       struct rxr_pkt_entry *pkt_entry)
+{
+	size_t hdr_size, data_size;
+	char * data;
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_pkt_type_req.c:879]: Spacing is off here.

File: prov/efa/src/rxr/rxr_pkt_type_req.c [hunk 6]
```diff
@@ -831,55 +899,121 @@ ssize_t rxr_pkt_proc_matched_medium_rtm(struct rxr_ep *ep,
...
+
+ssize_t rxr_pkt_proc_matched_read_rtm(struct rxr_ep *ep,
+				      struct rxr_rx_entry *rx_entry,
+				      struct rxr_pkt_entry *pkt_entry)
+{
+	struct rxr_read_rtm_base_hdr *rtm_hdr;
+	struct fi_rma_iov *read_iov;
+
+	rtm_hdr = rxr_get_read_rtm_base_hdr(pkt_entry->pkt);
+	read_iov = (struct fi_rma_iov *)((char *)pkt_entry->pkt + rxr_pkt_req_hdr_size(pkt_entry));
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_pkt_type_req.c:973]: This is kind of funky unpack logic in that it assumes `fi_rma_iov` is going to remain unchanged. That's probably the case, but regardless this shouldn't be in the wire protocol.

File: prov/efa/src/rxr/rxr_pkt_type_req.c [hunk 7]
```diff
@@ -992,6 +1132,16 @@ ssize_t rxr_pkt_proc_tagrtm(struct rxr_ep *ep,
 /*
  * proc() functions called by rxr_pkt_handle_recv_completion()
  */
+
+/*
+ * rxr_pkt_proc_rtm_rta() process a RTM and RTA packet entry
+ * that has been through the re-order buffer.
+ *
+ * It takes ep (end point) and pkt_entry as input
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_pkt_type_req.c:1140]: This is obvious from the function definition and doesn't add any information.

File: prov/efa/src/rxr/rxr_pkt_type_req.c [hunk 8]
```diff
@@ -1739,7 +1898,7 @@ int rxr_pkt_proc_write_rta(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
 	op = rta_hdr->atomic_op;
 	dt = rta_hdr->atomic_datatype;
 	dtsize = ofi_datatype_size(dt);
-	
+
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_pkt_type_req.c:1901]: Please remove unrelated changes.

File: prov/efa/src/rxr/rxr_pkt_type_req.c [hunk 9]
```diff
@@ -1384,6 +1534,10 @@ void rxr_pkt_handle_eager_rtw_recv(struct rxr_ep *ep,
 		return;
 	}
 
+	desc = rx_entry->desc[0];
+	ofi_copy_to_hmem_iov(desc ? desc->peer.iface : FI_HMEM_SYSTEM,
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_pkt_type_req.c:1538]: What if the iov in the write request is to a CUDA buffer? Doesn't this also need to be copied via fi_read?

**Review by rwespetal — CHANGES_REQUESTED** [review]
Please take another pass at the comments. Those need to be maintained too so if there are repeated comments or comments that are obvious from the code please remove them.

Also. not sure if the request to write also needs a local copy or if that's handled elsewhere.

File: prov/efa/src/rxr/rxr_pkt_type_data.c (hunk 4 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_data.c]: Removed in new revision.

**Review by wzamazon — COMMENTED** [review]
Thanks for your comments! I  just uploaded a new revision, which address your comments by removing excessive comments, and handle RTW types.

In the new revision, I did some refactoring: the logic that handle cuda/non-cuda was wrapped in function ```rxr_pkt_proc_data()``` and all packet handler call that function to copy data.

File: prov/efa/src/rxr/rxr_pkt_type_req.h [hunk 10]
```diff
@@ -329,6 +329,16 @@ void rxr_pkt_handle_read_rtm_send_completion(struct rxr_ep *ep,
 {
 }
 
+/*
+ * handle_data_copied() functions for RTM packet types.
+ * Note: read rtm does not contain data, so does not have such a handler.
+ */
+void rxr_pkt_handle_eager_rtm_data_copied(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry);
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_pkt_type_req.h:336]: This is duplicated/already on line 352

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_req.h:336]: Thanks! Fixed in new revision.

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
bot:aws:retest

**Review by rwespetal — APPROVED** [review]

File: prov/efa/src/rxr/rxr_pkt_cmd.c [hunk 11]
```diff
@@ -450,14 +452,155 @@ ssize_t rxr_pkt_wait_handshake(struct rxr_ep *ep, fi_addr_t addr, struct rxr_pee
...
+
+	pkt_entry->x_entry = rx_entry;
+
+	if (OFI_UNLIKELY(rx_entry->rxr_flags & RXR_RECV_CANCEL) ||
+	    data_offset >= rx_entry->cq_entry.len) {
+		rxr_pkt_handle_data_copied(ep, pkt_entry);
+		return;
+	}
+
+	if (efa_ep_is_cuda_mr(rx_entry->desc[0])) {
```
Comment (rajachan) [inline prov/efa/src/rxr/rxr_pkt_cmd.c:573]: Would this not fully avoid gdrcopy or whatever alternative copy implementation is available?

File: prov/efa/src/rxr/rxr_pkt_cmd.c [hunk 12]
```diff
@@ -450,14 +452,155 @@ ssize_t rxr_pkt_wait_handshake(struct rxr_ep *ep, fi_addr_t addr, struct rxr_pee
...
+		err = rxr_pkt_post_read_to_copy_data(ep, pkt_entry, data,
+						     data_size, rx_entry,
+						     data_offset);
+		if (OFI_UNLIKELY(err)) {
+			if (rxr_cq_handle_rx_error(ep, rx_entry, -FI_EINVAL))
+				assert(0 && "error writing error cq entry\n");
+		}
+	} else {
+		desc = rx_entry->desc[0];
+		ofi_copy_to_hmem_iov(desc ? desc->peer.iface : FI_HMEM_SYSTEM,
```
Comment (rajachan) [inline prov/efa/src/rxr/rxr_pkt_cmd.c:583]: The peer iface will always be FI_HMEM_SYSTEM here then. This conditional block is still a bit weird, ideally ofi_copy_to_hmem_iov takes care of the read-and-block logic so the provider does not have to care what type of MR this is, but I get why the current implementation is not setup to handle the async copy.

File: prov/efa/src/rxr/rxr_pkt_cmd.c (hunk 11 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_cmd.c:573]: gdrcopy is not used by EFA yet. We need a separate patch to enable gdrcopy for EFA. My plan is in that patch implement a function ```rxr_ep_use_local_read()``` to determine whether local read is used.

File: prov/efa/src/rxr/rxr_pkt_cmd.c (hunk 12 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_cmd.c]: Yes, ```peer.iface``` will always be ```FI_HMEM_SYSTEM``` with this patch.  I updated a new revision that call ```ofi_copy_to_iov``` instead of ```ofi_copy_to_hmem_iov```. 
I agree that ideally this conditional block should be inside ```ofi_copy_hmem_to_iov```, which would require some major change to that function.

**Review by rajachan — APPROVED** [review]
